src/index.ts

```diff
@@ -1,51 +1,69 @@
 import { compactDecrypt, importJWK, importPKCS8 } from 'jose'
-import {
-  Client,
-  ClientAuthMethod,
-  generators,
-  Issuer,
-  ResponseType,
-} from 'openid-client'
+import { Client, generators, Issuer } from 'openid-client'
 
+import {
+  API_VERSION,
+  DEFAULT_SCOPE,
+  DEFAULT_SGID_CODE_CHALLENGE_METHOD,
+  SGID_AUTH_METHOD,
+  SGID_SIGNING_ALG,
+  SGID_SUPPORTED_FLOWS,
+} from './constants'
 import { convertPkcs1ToPkcs8 } from './util'
 
-const SGID_SIGNING_ALG = 'RS256'
-const SGID_SUPPORTED_FLOWS: ResponseType[] = ['code']
-const SGID_AUTH_METHOD: ClientAuthMethod = 'client_secret_post'
+type ConstructorParams = {
+  clientId: string
+  clientSecret: string
+  privateKey: string
+  redirectUris?: string[]
+  hostname?: string
+}
+
+type AuthorizationUrlParams = {
+  state: string
+  scope?: string | string[]
+  nonce?: string | null
+  redirectUri?: string
+  codeChallenge: string
+  codeChallengeMethod?: 'plain' | 'S256'
+}
+
+type AuthorizationUrlReturn = { url: string; nonce?: string }
+
+type CallbackParams = {
+  code: string
+  nonce?: string | null
+  redirectUri?: string
+  codeVerifier?: string
+}
 
 export class SgidClient {
+  private apiVersion: number
   private privateKey: string
-
   private sgID: Client
 
   constructor({
     clientId,
     clientSecret,
     privateKey,
-    redirectUri,
+    redirectUris,
     hostname = 'https://api.id.gov.sg',
-    apiVersion = 1,
-  }: {
-    clientId: string
-    clientSecret: string
-    privateKey: string
-    redirectUri?: string
-    hostname?: string
-    apiVersion?: number
-  }) {
+  }: ConstructorParams) {
+    this.apiVersion = API_VERSION
+
     // TODO: Discover sgID issuer metadata via .well-known endpoint
     const { Client } = new Issuer({
       issuer: new URL(hostname).origin,
-      authorization_endpoint: `${hostname}/v${apiVersion}/oauth/authorize`,
-      token_endpoint: `${hostname}/v${apiVersion}/oauth/token`,
-      userinfo_endpoint: `${hostname}/v${apiVersion}/oauth/userinfo`,
+      authorization_endpoint: `${hostname}/v${this.apiVersion}/oauth/authorize`,
+      token_endpoint: `${hostname}/v${this.apiVersion}/oauth/token`,
+      userinfo_endpoint: `${hostname}/v${this.apiVersion}/oauth/userinfo`,
       jwks_uri: `${new URL(hostname).origin}/.well-known/jwks.json`,
     })
 
     this.sgID = new Client({
       client_id: clientId,
       client_secret: clientSecret,
-      redirect_uris: redirectUri ? [redirectUri] : undefined,
+      redirect_uris: redirectUris,
```

Comment:
Just wanna say good job for fixing this! Courtesy of @PrawiraGenestonlia also of course